llvm/test/Transforms/IndVarSimplify/casted-trunc.ll

```diff
@@ -0,0 +1,28 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt < %s -passes=indvars -S | FileCheck %s
+
+declare void @foo(i16 noundef)
+
+; Function Attrs: mustprogress noreturn uwtable
+define void @bar(i64 noundef %ptr) {
+; CHECK-LABEL: @bar(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[TMP0:%.*]] = trunc i64 [[PTR:%.*]] to i4
+; CHECK-NEXT:    [[TMP1:%.*]] = zext i4 [[TMP0]] to i16
+; CHECK-NEXT:    br label [[WHILE_BODY:%.*]]
+; CHECK:       while.body:
+; CHECK-NEXT:    tail call void @foo(i16 noundef signext [[TMP1]])
+; CHECK-NEXT:    br label [[WHILE_BODY]]
+;
+entry:
+  br label %while.body
+
+while.body:                                       ; preds = %entry, %while.body
+  %0 = phi i64 [ %ptr, %entry ], [ %add.ptr, %while.body ]
+  %1 = trunc i64 %0 to i16
+  %and = and i16 %1, 15                           ; loop invariant
+  tail call void @foo(i16 noundef signext %and)
```

Comment:
trunc may affact the result, so may be need a negative test, similar to `%1 = trunc i64 %0 to i2` ?

Reply:
Having something like:

`declare void @foo(i2 noundef)

entry:
  %ptr.addr.0 = ptrtoint ptr %ptr to i64
  br label %while.body

while.body:                                       ; preds = %entry, %while.body
  %0 = phi i64 [ %ptr.addr.0, %entry ], [ %add.ptr, %while.body ]
  %1 = trunc i64 %0 to i2
  %and = and i2 %1, 15                           ; loop invariant
  tail call void @foo(i2 noundef signext %and)
  %add.ptr = add nsw i64 %0,  16
  br label %while.body
}`

With this change, the IR generated is the following:

`declare void @foo(i2 noundef)

define void @bar(ptr noundef %ptr) {
entry:
  %ptr.addr.0 = ptrtoint ptr %ptr to i64
  %0 = trunc i64 %ptr.addr.0 to i2
  br label %while.body

while.body:                                       ; preds = %while.body, %entry
  %and = and i2 %0, -1
  tail call void @foo(i2 noundef signext %and)
  br label %while.body
}`

This is valid to me, when you say that `trunc` can affect the result, do you have any example in mind? Thanks!

Reply:
What I've also observed from the regression is that we have one test failing: `test_11` in `Transforms/IndVarSimplify/X86/eliminate-trunc.ll`.

In concrete, in the function the statement failing is:

```ll
define void @test_11() {
...
bb5:                                              ; preds = %bb5, %bb5
  %_tmp24 = icmp slt i16 %_tmp15, 0
  br i1 %_tmp24, label %bb5, label %bb5
...
}
```

which instead of the expected (as it is currently in the test):

```ll
; CHECK:       bb5:
; CHECK-NEXT:    [[_TMP24:%.*]] = icmp slt i16 poison, 0
; CHECK-NEXT:    br i1 [[_TMP24]], label [[BB5:%.*]], label [[BB5]]
```

It gets simplified to:

```ll
bb5:                                              ; preds = %bb5, %bb5
  br i1 poison, label %bb5, label %bb5
```

I'm assuming this simplification is correct, but I don't understand why this is happening with this change.

Reply:
Oh, sorry, ignore this. 
yes, it is correct.  https://alive2.llvm.org/ce/z/Esz3Vt